clientv3/integration/network_partition_test.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2017 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !cluster_proxy
+
+package integration
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/integration"
+	"github.com/coreos/etcd/pkg/testutil"
+)
+
+// TestNetworkPartitionBalancerWithPut tests when one member becomes isolated,
+// first Put requests fail and following retry succeeds with client balancer
+// switching to others.
+func TestNetworkPartitionBalancerWithPut(t *testing.T) {
+	testNetworkPartitionBalancer(t, func(cli *clientv3.Client) {
+		var err error
+		for i := 0; i < 2; i++ {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			_, err = cli.Put(ctx, "a", "b")
+			cancel()
+			if err == nil {
+				break
+			}
+			if err != context.DeadlineExceeded {
+				t.Fatalf("#%d: expected %v, got %v", i, context.DeadlineExceeded, err)
+			}
+			if i == 0 { // give enough time for endpoint switch
+				time.Sleep(5 * time.Second)
```

Comment:
we need to get rid of this eventually. probably need a wait endpoint switch func.

Reply:
We can expose `balancer` interface with `ConnectNotify`. Should we?

Reply:
for this pr. simply add a todo is good enough.